src/FakeItEasy/IoC/DictionaryContainer.cs

```diff
@@ -8,120 +8,69 @@ namespace FakeItEasy.IoC
+        private object Resolve(Type serviceType)
         {
-            private SingletonResolverState state;
-
-            public SingletonResolver(Func<DictionaryContainer, T> resolveFunction)
-            {
-                this.state = new UnresolvedState(this, resolveFunction);
-            }
-
-            public T Resolve(DictionaryContainer container)
+            if (this.cachedServices.TryGetValue(serviceType, out object cachedService))
             {
-                return this.state.Resolve(container);
+                return cachedService;
             }
 
-            private class ResolvedState
-                : SingletonResolverState
+            if (this.serviceFactories.TryGetValue(serviceType, out Func<DictionaryContainer, object> serviceFactory))
             {
-                private readonly T instance;
-
-                public ResolvedState(T instance)
-                {
-                    this.instance = instance;
-                }
-
-                public override T Resolve(DictionaryContainer container)
-                {
-                    return this.instance;
-                }
+                var serviceObject = serviceFactory.Invoke(this);
+                this.cachedServices[serviceType] = serviceObject;
+                return serviceObject;
             }
 
-            private abstract class SingletonResolverState
-            {
-                public abstract T Resolve(DictionaryContainer container);
-            }
-
-            private class UnresolvedState
-                : SingletonResolverState
-            {
-                private readonly SingletonResolver<T> resolver;
-                private Func<DictionaryContainer, T> resolveFunction;
-
-                public UnresolvedState(SingletonResolver<T> resolver, Func<DictionaryContainer, T> resolveFunction)
-                {
-                    this.resolveFunction = resolveFunction;
-                    this.resolver = resolver;
-                }
-
-                private bool SingletonHasNotBeenCreated => this.resolveFunction != null;
-
-                public override T Resolve(DictionaryContainer container)
-                {
-                    lock (this)
-                    {
-                        if (this.SingletonHasNotBeenCreated)
-                        {
-                            var instance = this.resolveFunction(container);
-                            this.resolver.state = new ResolvedState(instance);
-                            this.SignalThatSingletonHasBeenCreated();
-                        }
-
-                        return this.resolver.Resolve(container);
-                    }
-                }
-
-                private void SignalThatSingletonHasBeenCreated()
-                {
-                    this.resolveFunction = null;
-                }
-            }
+            throw new KeyNotFoundException($"The specified service {serviceType} was not registered in the container.");

```

Comment:
Shouldn't there be some locking in this method? If two threads attempt to modified `cachedServices` at the same time, it could become corrupted. Or maybe use a `ConcurrentDictionary`?

Reply:
And... ignore this too. All services are constructed eagerly from a single thread, and the dictionary won't be modified afterward.

Reply:
It's currently safe, since we're only accessing the `DictionaryContainer` from the static constructor of `ServiceLocator`, which is guaranteed to run once. If you want futureproofing, I can add.